### ISSUE TITLE
Thank you for effort, but ...

### ISSUE THREAD

**Issue by elgreko5** [aaFn/Bookmark-search-plus-2#89]
Hello aaFn,

First, I am really grateful for your effort (I am sure many others are, too), but I also think you're going in the wrong direction.

This will be a bit long-winded, but please bear with me.

Bookmarks are absolutely essential part of a browser, and FF not having functionality like in Show & Go parent folder today not functioning add-ons on its own, is a sign of idiocy of FF developers. As is e.g., moving FF Home Page icon from right to left while not rectifying a bug which caused losing your session tabs (at times I had more than 1.200 of them; after futile attempts to find a solution online, I found it myself - replace "_closedW" with "w" in sessionstore.js), not rectifying a bug which caused losing years of browsing history and shrinking it to a week, etc.

Bookmarks, tabs, browsing history, are very important data (I have about 80.000 of bookmarks and some are 15 years or so old), and should be treated as such. Imagine just the time needed to bookmark 80.000 links. Now imagine the time to open 80.000 pages.
Imagine the time to find, read, select, etc. and finally bookmark 80.000 pages... And that's just about invested time.

So, bookmarks database must be safe. If I understood correctly you made the effort to safeguard favicons, but not bookmarks. Also, I see quite a bit of options for favicons. And favicons are nowhere near as important as bookmarks, if at all?! Favicons are a little more than eye-candy, and candy is for kids. (Also, when there are no loaded favicons, or they are disabled, BSP displays a blotch - it should be something neutral or nothing at all.)
So, safety first.
Regarding safety of data - earlier today, your add-on's statistics showed more than 80.000 bookmarks - now it shows 79.672. I assure you I didn't delete 500 bookmarks today. (?!)

Efficacy second.
3-4 seconds wait to "Load saved state" or "" (yes, sometimes there is a message and sometimes there is not - wait is always there) on an i5@3.2-3.7 8GB SSD computer is ridiculous. 
Sorry, but it really is. Long time ago, when 486 or 386 was a computer, I wrote a program for POS that replaced "program" that a company was using that needed 20 or more minutes just to display information about available quantity of goods for a customer. My program gave that info instantaneously and I didn't do anything exceptional - the solution was trivial. In fact, there was nothing to solve, it was just that my predecessor didn't have a clue.
3-4 s wait when clicking on "Bookmarks menu" to close it, too. I don't know if that is your fault or FF's - if it is FF's and can't be resolved, it is time for me to finally stop using it.
Outside of your add-on, said opening and closing is almost instantaneous.

Information design - user interface
Possibility of changing appearance of bookmarks menu in your add-on is good, but forcing some unknown set of font size and item spacing and whatnot (it seems the font is changed too - I lost my leading spaces indentation) a new default is not!

Also, when right-clicking on a bookmark, I do not get context menu anymore (I did yesterday or day before - I think I then was on a "beta update channel")?!

Regarding my 80.000 bookmarks - it seem there are unique 40.000 of them. At one time, when I was importing bookmarks from IE, idiotic FF duplicated them. Since then I have two Bookmarks Menus - second is in Other Bookmarks. Problem is that those bookmarks in that second Bookmarks Menu are updated automatically as I edit them in original Bookmarks menu. EDIT: Checked, there are 80.000 bookmarks, only folder was duplicated.
This doesn't have anything to do with your add-on, but what does is that when BSP is active, instead of second "instance" of Bookmarks Menu in Other Bookmarks, now there is only a settings icon without any text beside it?!

![1](https://user-images.githubusercontent.com/47822770/53136270-b2987c80-357e-11e9-841c-8774798475d1.gif)
![2](https://user-images.githubusercontent.com/47822770/53136283-b926f400-357e-11e9-9be8-5d54571e9834.gif)

Also, today after I edited one bookmark and searched for it through BSP, BSP showed name like it was before editing?! Disabling and enabling BSP didn't help - situation resolved only after restarting FF.

One more suggestion - it would be pretty useful if you could add counter of tabs to your add-on.

So, thank you for effort, but there is much room for improvements, some of those - essential.

Kind regards

**Comment by aaFn** [comment]
Hello @elgreko5 ,

thank you for this extensive feedback
Quickly:

1. FF flawed approach
Agree .. Purpose of this add-on is to disappear when FF incorporates the function. Not going to happen tomorrow I am afraid, given the duration this demand has been put on FF (basically since the beginning .. https://bugzilla.mozilla.org/show_bug.cgi?id=196509)

2. Bookmarks database
Bookmarks  are stored in Firefox (FF), not in the add-on. The add-on only have a cache of them (because of FF API limitations :-( ), but really the add-on main function is to display bookmarks and allow showing where they are

3. Safety of bookmarks
If you lost 500 bookmarks in stats, this is due to FF. BSP2 as I said does not store them, only displays them. Only when you do an action like "delete" on BSP2 can you really remove something.

4. Favicons
I didn't make efforts on favicon by choice, but because of https://bugzilla.mozilla.org/show_bug.cgi?id=1315616
Basically, FF API doesn't give any add-on access to the favicon DB (contrary to Chrome) ... so I had to develop something to safely get them (no security exposure, no data privacy leak ..) by directly fetching them in background from the sites themselves. Meanwhile, an icon with a little blue T at bottom right to mean this is a temporary icon waiting for the fetch to occur, is displayed.
You can turn this off in the options, and a standard "nofavicon" icon is displayed immediately.

5. Delay to load / efficiency
A looong story .. again related to FF API limitation, which got even worse recently.
See FF bug -> https://bugzilla.mozilla.org/show_bug.cgi?id=1396364
and other details in https://github.com/aaFn/Bookmark-search-plus-2/issues/2
The origin is awful coding and data mixing inside FF (I could cast an eye at it ...), with tags handled as bookmarks ..
At the costs of many efforts, with building a local cache, I could accelerate all subsequent loads, after the initial one, by an exponential factor depending on the number of bookmarks .. but not the initial one
The value I am at today is as far as i could get today with using FF API on local storage (which is far from efficient, it is storing json text, not binary, and it is slooow and hungry in memory..) and on the HTML display speed on FF, and I believe I got all the juice I could get from it.
All the rest is marginal or so complex to code I do not want to enter into it, given the purpose of the add-on is to disappear one day when FF realizes the function.

6. Native bookmarks speed
This is apparently using native OS primitives, or it has access to things add-ons cannot access, so display and storage access are much more speedy ... cannot do much about it, FF API is very **limited.**

7. Bookmarks menu speed to close
Sorry, this is native FF and it speed of HTML display / manipulation, not much I can do about .. I already tried to optimize the HTML used to display to the simplest possible form fitting with needed features.

8. Right click context menu
Should work. You might have memory issues though, FF has a memory hungry monster inside it, and it appears to be triggered when fetching favicons.
Try disabling favicon fetching in the options.

9. Leading spaces
You appear to be the second one saying that .. could you be more specific (a screenshot ?), I tested, this is displayed and working on my Windows or Linux machines.

10. Duplicate Bookmarks Menu inside Other Bookmarks
Looks like this is something special ... what do you get when you hover on the "Setting" favicon ?
Can you look at its properties ?
Normally, this favicon corresponds to the "Most visited" bookmark item, or the "Recent Tags" one.
I presume something is corrupted in your bookmarks structure.

11. Edit of bookmark not retrieved immediately
Did you do that in BSP2, or in the native bookmarks toolbar ?
It the first, it is possible that the update to the native DB got delayed in its notifiaction back to BSP2 display front end, because of FF memory state or cpu load.
Again, try disabling favicon fetching in case this is due to the memory monster ...

12. Count of tabs
Do you mean count of open tabs on screen ?
I have been toying with the idea of incorporating tabs search and handling in the add-on since the beginning, but for now decided to stay focused on bookmarks.
Maybe one day I will add tabs indeed ..; I didn't mature enough on that decision yet.

Let me know if you have any additional feedback,
kind regards, aaFn.

**Comment by elgreko5** [comment]
Hello aaFn,

Thank you for your prompt response!

1. I am sure your add-on will be needed for years to come. The fact its been 16+ years and that they are writing garbage like - they lack the resources (16 y no resources for show and go folder?!) or that 95+% of their users have <2.000 bookmarks (like 2.000 is 20) assures me of that. If they, God forbid, wrote Total Commander, we would be able to find and open files, but not to know where they are and wouldn't be able to copy them...

2. I knew that, of course.

3. Of course, but:
![1](https://user-images.githubusercontent.com/47822770/53172151-ed350000-35e4-11e9-92e1-951c26baa485.gif)
I know I saw BSP2 displaying more than 80.000 so it must be it's fault.

4. Knew that also. Maybe I wasn't clear - I think that favicons are the least important thing in your add-on. Again, they are next to useless. What's more, in a case like mine, I would be watching fetching them for years, and we don't need another memory hog provoking thing in FF...

5. If I understand you correctly - now even alice0775's Show & Go parent folder would have to have that delay?

8. No memory issues, I would have spotted them:
![4](https://user-images.githubusercontent.com/47822770/53175657-bc58c900-35ec-11e9-9438-08eab6768f84.gif)
Already disabled favicons yesterday because of uselessness.

9. 
![2](https://user-images.githubusercontent.com/47822770/53175985-6f292700-35ed-11e9-9284-fcd1a5b9cee8.gif)
![3](https://user-images.githubusercontent.com/47822770/53175997-751f0800-35ed-11e9-80e8-5388e29487c9.gif)
Font has changed. Also, icons are not what you described but what I called blotches.

10. Hovering gets me nothing. Right click doesn't get me properties but "New bookmark, New folder, New separator, Paste info" menu - the same as when I right click on Other Bookmarks.

11. Don't remember that, but I am pretty sure it wasn't memory or CPU - I would have noticed that.

12. Yes, I meant that.

Right click context menu still doesn't work.
Tried on another computer which never had BSP2 installed - doesn't work.
As I said, it seems it works on beta versions.

Kind regards

**Comment by elgreko5** [comment]
Another bug in BSP2: 
Sort By Name puts "1" before "_"

Also, it is so slow you can actually see as bookmarks are printed in new order.
There is also another anomaly (obviously caused by slowness) which I am not sure I can explain exactly but it looks like if you do Sort By Name and immediately scroll down and click on a bookmark, sidebar bookmark list jumps.

**Comment by aaFn** [comment]
Hello @elgreko5,

3. Indeed, BSP2 only counts bookmark items (= bookmarks, folders and separators), as it is using the bookmarks interface https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks

4. Yes agree, this is a cosmetic thing, although many people like them because this helps visually find bookmarks, as an acceleration.
Still cosmetic though, this is why I made an option to disable them.

5. In its time, alice0775's extensions (not add-ons, note the difference) were directly integrated within the FF interface, thanks to XUL technology. This is not the case anymore with the WebExtension API https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions, add-ons are totally external and can only interact with FF and the user through these API's and through HTML.
So alice0775's extensions could be fast, as relying on FF interface itself and its OS level primitives.
Not possible anymore nowadays ..

6.
7.

8. Re-reading you, I may be misinterpreting what you call "Right click context menu".
Are you talking of the context menu inside BSP2 sidebar itself, or of the BSP2 submenu addition which is giving the path to bookmark as a string ?
If the second, this is normal .. FF currently doesn't support that menu extension on the native Bookmarks sidebar, nor within the Bookmarks Library, as per -> https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus/ContextType
I am glad to hear that this is working on beta, that means this is coming :-) (FF people are progressing after all ...)

9. Ok, I spotted the problem .. it appears HTML is reducing all the leading spaces to only one space character.
I will see if I can oblige it to respect all characters, and not condense spaces to only 1.

10. BSP2 only gets what this API call gives to it https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/getTree, without interpreting it.
I will check nothing is misinterpreted if a bookmark id is seen twice in the tree (they are supposed to be unique), but I do not think so, and in that case there's nothing I will be able to do.
I'll keep you posted.

11. Ok, not sure what is the cause of delay. Did you make the edit from inside BSP2 using the "Properties" dialog, or from the native bookmarks sidebar (or Library) ?

**Comment by elgreko5** [comment]
4. Yes, they can be useful (they are really useful in tabs), it's just that I thought there are still wrinkles that should have been ironed out before spending time on favicons - e.g. changed fonts

8. Seems you skipped last paragraph - excerpt: "Yesterday I thought it worked in sidebar, but it obviously didn't. On beta it works in Library and Bookmarks drop-down menu, but on regular it only works in menu (where it is useless, don't understand why you even put it there).
I read you wrote that it can't work in sidebar because of FF limitations."

9. No, as I already wrote - font has changed (as well as interlinear spacing) - it is obvious from the screenshots. Number of spaces is the same - three - it's because of different proportionality of different fonts. Also, look at the favicons - blotches.
Correcting this, if you ask me, should be a priority. Also, it seems it would be trivial.

10. Ok, that's why I reported it to you. I will eliminate that clone folder eventually, it's just that I'm not really looking forward to that endeavor...

11. Can't be sure, I'm afraid. Hasn't happened since.

**Comment by aaFn** [comment]
Ok, thank you for the quick return.

On the total number of items, on my side, this matches perfectly = (number of bookmarks + number of oflders + number of separators) as counted by BSP2 exactly matches the counts in the import menu in the Library.
So I am not sure I can tell what is wrong in your case, but the double presence of the same tree / objects under different places could impact such counts maybe.

Changed fonts = I do not know which font is used by FF natively, they vary by OS and by version, so I have to try to find them manually (if available to the add-on):
- On all my Windows 10 machines (physical and virtual), BSP2 visually has the same font than the native Bookmark sidebar so I believe I found the right one there.
- On my Linux with Mate and KDE desktops, same (but by increasing the BSP2 font size to 14), so should be ok.
- Can't tell about IOS, I have no such machine.
- On Windows 7, the font is different .. It appears Windows 7 is using Tahoma, and that font is not availble to the add-on (or was not last time I tried). Plus, windows is Windows, hard to make any difference from within the add-on between Win 7 and Win 10 .. so I favored Win 10, which is more numerous now.
So I confirm, things look different on Windows 7 (which appears to be your case from screenshots).

I do confirm also the collpasing of multiple spaces into one by HTML. I found the way to prevent that by using `white-space: pre;` in the CSS. This improvement will be in 2.0.54.

**Now, on the other bugs you mention;**

1. Sort by name
This is using ASCII code, and using the Javascript native '<' operator.
See http://www.asciitable.com/ : `1` character code is 49 (decimal), and `_` character code is 95 (decimal)
So it is normal that `1` is put before `_`.
I do not know what character ordering the native bookmark is using, but that seems non natural. If there is a simple logic I can reproduce, ok to try to change, else I will remain with ASCII and Javascript sorting order.

2. Slowness = each bookmark move is sent to FF by this API call https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/move. 
There is no way to send all or several moves in a bunch.
Then, slowness comes for a part from this "one-by-one" send and FF API reactivity to the calls.
In all truth, this is not slow with 2000 bookmarks .. but for a second part, things also slow down with high numbers of bookmarks, because the whole list of bookmarks in BSP2 is an HTML table and not an OS level list. With high number of rows, HTML gets slower in individual delete / insert operations (there is no row move operation on HTML tables).

3. Jumps = not really surprised by that, since bookmarks continue to move .. this is impacting where is the cursor I guess, or making things jerky on the screen movements because there are HTML refreshes in the middle (related with slowness / cpu intensive operations on the HTML table as said above).

4. Properties window
This is an FF / Chrome security choice = Modal OS dialogs are not possible anymore.
They can only be browser windows ... and so such windows can remain open when other windows are closed, like any other browser window.
And of course, FF restores them when reopening, like any other browsing Window ...
Can't do anything about it, sorry :-( .. talk with Chrome and FF about their choices for our good ... (some of those choices look like the Gnome 3 diktats, with these developers know better than us what is good and intelligent for us .. as a developer myself, I call that self-justified developer lazyness instead of intelligence)

5. Auto scrolls when dragging
A complex handling. There is no way to disable or to control auto scrolling on drag. I tried and spent many hours on that, I can tell you .. and found FF bugs in the process btw !
At the end, I resorted to a trick of removing scrollbars in some specific cases, when I am able to detect that a drag is coming from outside to inside, but once we come to deeper inside the sidebar, I have to restore the scrollbars, so as to allow the user to use natural auto-scrolling for placing the bookmark at the right place ...
So what you observe is normal, and wishable since it allows somebody to auto-scroll where he/she wants to go (imagine a very deep tree, with bookmarks out of sight on right of the sidebar window .. need to auto scroll there on drag to be able to drop at the right place .. something not possible with the native bookmark sidebar by the way, since it doesn't have any horizontal scrollbar).

Hope that answers all the points you raised.
I am still looking details on item 10) to verify if there's something I can do or correct.

**Comment by aaFn** [comment]
Ah yes, forgot abouth what you call the "blotches".
This was the favicon used by the native Bookmark sidebar in FF 57 for bookmarks without fetchable favicons.
I just captured the screen and made it my own "nofavicon" icon.

It has changed now, probably around FF 62 (and I guess you could still call it a "blotch" btw .. myself, I don't like the new any better than the previous one ... both look ugly to me), so I will recapture it, and refresh in 2.0.54.

**Comment by aaFn** [comment]
So looking at 10) .. I believe the duplicate "Bookmarks menu" is represented as a "place:" URL bookmark.

These are special bookmarks for which I have to look at the "sort=xx" part in the URL to identify what they are (if xx = 8, then this is "Most Visited", if xx =14 then is this is Recent Tags and if xx =12 then this is Recent Bookmarks).
And then I have to emulate them, because the FF API does not return to me their content, and add-ons do not have access to "place:" now ("place:" appear to be live queries of sets of bookmarks and other objects, executed by the native Bookmark sidebar directly on sqlite DB .. kin of like a "View").

If I am correct, this case is too specific and rare, I won't enter into coding things to emulate it.
What I'll do is just show the URL as its title, instead of having no title, so that it is more obvious that it is a special thing, and also to verify my hypothesis ..
To be in 2.0.54.

**Comment by elgreko5** [comment]
I finally took to resolve the clone folder situation.
With situation being that when a bookmark is deleted from a clone folder, it is also deleted from original folder, I presumed deletion of a clone folder would delete original folder too. So I delved into bookmarks.json and manually excavated this:
BOF
,"children":[{"guid":"ho1czFgP8sJX","title":"","index":0,"dateAdded":1320785492187000,"lastModified":1550759278609000,"id":31850,"typeCode":1,"annos":[{"name":"places/excludeFromBackup","value":"1","expires":4,"flags":0}],"type":"text/x-moz-place","uri":"place:parent=menu________"}]"
EOF
The part immediately before removed text is this:
BOF
{"guid":"unfiled_____","title":"Other Bookmarks","index":3,"dateAdded":1313873732468000,"lastModified":1550971885685000,"id":5,"typeCode":2,"type":"text/x-moz-place-container","root":"unfiledBookmarksFolder"
EOF
After removal I was left with a .json file of 39,548,576 bytes.
Then I deleted the offending folder just to see if original folder will be deleted too, and, lo and behold - it wasn't! Did a backup of bookmarks and guess what - resulting file had a length of 39,548,576 bytes?!
Being used to so much more pain from Mozilla idiots, I couldn't believe it - I wasted only 20 minutes?! 
Library's Restore option now shows 81949 items, while BSP2 shows 81947 (79771+2169+7+0), so it seems OK now. (On another W7 PC (with 165 bookmarks) there is also a discrepancy of 2, as well as on another W10 PC (17 bookmarks).)

> Changed fonts

Yes, using W7 on this machine. I overrode interlinear spacing to 6 and it is much better now, although the font is uglier than original and, more importantly, less legible - it looks like condensed print on a 9-pin dot matrix printer.
Maybe you could ask Mozilla which is the original font or try Windows 7 system font - Segoe UI (it sure looks like it) and then make it an option? Just looked it up - W10 - 53%, W7 - 35%...

>     1. Sort by name
>        See http://www.asciitable.com/ : `1` character code is 49 (decimal), and `_` character code is 95 (decimal)

Yes, I know that (for more years than I care to admit :) )

>        I do not know what character ordering the native bookmark is using, but that seems non natural. If there is a simple logic I can reproduce, ok to try to change, else I will remain with ASCII and Javascript sorting order.

In fact, natural sorting exists and it is not ASCII sorting. Also, sorting in BSP2 is not ASCII sorting, e.g. it puts "_" before "A" but after "1".
But, that's not the point here - the point is that BSP2 sorting is different from original and leaving it like this is inconsistency.

>     2. Slowness
>        In all truth, this is not slow with 2000 bookmarks

Please don't get me wrong when I say - and if it was slow with 2000 bookmarks, it could be said - but it isn't with 20... I understand that there is nothing more you can do.

>     4. Properties window
>        They can only be browser windows ... and so such windows can remain open when other windows are closed, like any other browser window.

Wow! Mozilla Geniuses at work!

>        Can't do anything about it, sorry :-( .. talk with Chrome and FF about their choices for our good ... (some of those choices look like the Gnome 3 diktats, with these developers know better than us what is good and intelligent for us .. as a developer myself, I call that self-justified developer lazyness instead of intelligence)

Yes, but that's not just laziness, it is already a pathological state, so, no, I won't be talking to them :)

>     5. Auto scrolls when dragging
>         but once we come to deeper inside the sidebar,
>        So what you observe is normal, and wishable

It seems you didn't understand - it happens even when you have JUST crossed the vertical scrollbar (into the sidebar) and moved back across it. If you do the same SHORT movement again, it scrolls more.

> Hope that answers all the points you raised.

It seems you skipped this: "Clicking on a light blue space beyond icon and text doesn't open folder although it should.". There is a screenshot above.

> "nofavicon" icon.

If it's not too much to ask, could you just replace it with something neutral - less distracting.

Kind regards

**Comment by elgreko5** [comment]
Another inconsistency:
Open in a New Tab in Bookmarks opens a new tab as last of opened tabs.
The command with a same name in BSP2 works differently - opens a new tab after current tab.
This is even better but name should be changed to reflect difference. Also, I think the best solution would be to have borh command in BSP2.

**Comment by aaFn** [comment]
Interesting ideas. And by the way, thanks for all the testing :-)

----------------------------------
You are correct, from what I know, the redundant folder you had was not really a folder, but a view at the real folder.
Deleting it had no effect on the real one, it just deleted the view at it.
However, deleting an object inside was in fact deleting the true object, because this was a view.

----------------------------------
On fonts, yes I do set Segoe on Windows, whatever version it is:
```
.fontwin {
  font: 12px "Segoe UI";
}
```

What you see in Windows 7 is the result of that setting.
Btw, if you activate traces in the options, you will see that under Windows 7. It is a checking that I make afterwards to see how FF transformed what I have set:
```
...
PlatformOs: win
Setting Windows variations
...
fontFamily = '"Segoe UI"'
fontSize   = '12px'
...
```
So it does not look like the native Bookmark sidebar is "Segoe UI" since we see it is different.
I do believe it is "Tahoma", but last time I checked I cannot access that font from the add-on.
What that means also is that Sego UI doesn not look the under Windows 7 and Windows 10 ...

If you want to have a closer look, go to C;\Windows\Fonts, and look at both "Segoe UI" and "Tahoma" .. I believe you should conclude like me that this is Tahoma they are using, but once again I can't ...

----------------------------------
Sort by name ..
Something I didn't say = like the native Bookmark sidebar, I am sorting with ignoring caps.
So All Caps are first converted to Small letters, and then sorting is done.
This is why `_` is before `A` .. =because in fact it is before `a`

I still don't know what the native Bookmarks sidebar is using as sort order, but I believe I am not too far from it.

You are correct though that as long as it is not exactly the same this is inconsistent. If we can find what it is, then I'll be ok to switch to it ...

----------------------------------
Horizontal scrolling when crossing the vertical separation.

Well, I do not have that effect on any of my machines (win10, Win7 or Linux).
Here is what is designed.
It happens exactly like that for me, and so should happen for you also (plenty of hours of tuning to get that ... because of FF bugs and insufficient API to manage these things ..):
- Take the (i) icon in the address bar or anything other to drag, from outside the sidebar
- Drag by the right of the sidebar = when crossing the sidebar limit, an interdiction symbol appears as long as you are over the vertical scrollbar
- Continue towards left, as soon as you leave the vertical scrollbar, both scrollbars disappear (horizontal and vertical), and (normally) no scrolling happen.
I am saying "normally", because in fact a scroll does happen, I cannot avoid it :-(, but I am catching it and forcing the frame back to its last remembered position before the scroll .. not visible "normally" because very quick ..
And I cannot avoid it because I cannot control the auto-scroll area size and position ..
- At that point if you go back to right, nothing happens until you get out of the sidebar, at which time the scrollbars are restored after a timeout (another trick .. a bug in FF is removing one key event supposed to notify me that the drag exited ... so I have to rely on interruption of the supposed_regular (another bug ..) mouseover event to detect the mouse is no more on the sidebar and then I can restore the scrollbars.
- If you go deeper left instead, after some distance the scrollbars are restored to allow the wishable auto-scrolling function

Apparently, this is not what is happening on your side.

Are you able to describe exactly in the same terms what is happening to you ?
And would you be so kind as to open the Browser console before doing the drag test (Ctrl+Shit+J), empty it with the trash icon at top left, and let me know if you see any message / error happening there while you do the experiment ?

----------------------------------
Clink on space before (or after) the "favicon + text" ... sorry I  missed this point.

Correct, that does nothing by design, see -> https://github.com/aaFn/Bookmark-search-plus-2/wiki/Mouse-clicks-and-modifiers-summary-table-(as-of-2.0.46-)
I find this is an inconsistency in the native Bookmark sidebar = when you click on the space at left of a bookmark, nothing happens. So why should it when it is a folder ?

Doing so allows to select a folder like any other item, without any action, and then with the keyboard up/down arrows you can go up/down.
On the native Bookmark, to do the same, you have to click twice on the folder to be able to navigate with keys after that ... very cumbersome :-(

So in short, if you want an action, click on favicon or text. If you click outside of that, no action, apart from setting the cursor at that point.

----------------------------------
Nofavicon icon = in 2.0.54, I am replacing by the new "nofavicon icon" of the latest FF version. Hope this is less distracting. At least this is aligning with their icon design change.

----------------------------------
Open in new tab = have both forms of next and last

Indeed, that is an idea. However, mind that depending on an option of FF, open in new tab can be the next tab instead of the last tab .. so we cannot say that "Open in new tab" is always open as last.

What (short please) labels would you suggest for open next and open last ?

----------------------------------
Thank you, aaFn

**Comment by elgreko5** [comment]
> Interesting ideas. And by the way, thanks for all the testing :-)

Glad to help.

> On fonts, yes I do set Segoe on Windows, whatever version it is:

You're right for Segoe UI and Tahoma. To be absolutely sure I am seeing the same, I checked with the help of layout.css.devPixelsPerPx=3.
Could you try with Arial - looks like a good substitute. I think it might be even better because of a tad bigger inter-character spacing than that of Tahoma.
 
> Sort by name ..

Finding out the system would be easy, if not a little boring. Will do it when I catch some time.

> Horizontal scrolling when crossing the vertical separation.

Tried on a W10 - the same thing happens.
On a W7 caught only this: TypeError: event.originalTarget.getAttribute is not a function[Learn More] tabbrowser.xml:1588:30

> (plenty of hours of tuning to get that ... because of FF bugs and insufficient API to manage these things ..):

I know exactly what are you talking about - had my share of building UIs from scratch because standard UIs weren't neither ergonomic nor efficient. Would like to put that fella Marco from M. through the steps, bet he'd change his tune.

> Are you able to describe exactly in the same terms what is happening to you ?
> And would you be so kind as to open the Browser console before doing the drag test (Ctrl+Shit+J), empty it with the trash icon at top left, and let me know if you see any message / error happening there while you do the experiment ?

Tried, just one error (above). When I catch some time will do, or maybe better - make a video.

> I find this is an inconsistency in the native Bookmark sidebar = when you click on the space at left of a bookmark, nothing happens. So why should it when it is a folder ?

I agree, it is. 

> Doing so allows to select a folder like any other item, without any action, and then with the keyboard up/down arrows you can go up/down.
> On the native Bookmark, to do the same, you have to click twice on the folder to be able to navigate with keys after that ... very cumbersome :-(
> So in short, if you want an action, click on favicon or text. If you click outside of that, no action, apart from setting the cursor at that point.

However, I think this would be even better: clicking on a left space selects item - folder or a bookmark, but clicking on the right space opens item. Notice that opening an item is far more used operation than selecting and right space opening is reducing mouse mileage. Also, it'd be consistent with FF Bookmarks thus eliminating frustrating failed attempts at clicking (which would result in selecting). Also, often there is no right space next to a bookmark, especially if it's in a subfolder, hence no chance of selecting. 

While at it, I read what you wrote about multi-selection and that you postponed it's implementation, but - would it be easier to implement it first to work with keyboard? 

> What (short please) labels would you suggest for open next and open last ?

Open in a New Next Tab
Open in a New Last Tab

I see you shortened it in BSP2 to "Open in new tab", then above without an "a" :)

**Comment by aaFn** [comment]
Unfortunately, as said previously, I cannot distinguish at 100% and in a durable way betweein Windows 10 and Windows 7. This is Windows and that's all.
So I am favoring Windows 10, by setting the correct one for Windows 10.

For horizontal auto-scrolling, the error you catch is not coming from BSP2 .. not sure what it means or if this is related to what is happening.
I'll wait that are you are able to edscribe what is happening in more details then to see if I can figure out what could be happening.

Currently, click on right or left of favicon+ text is the same as seen from the code, with no easy way to distinguish. I'll see if there is a more obvious way.

Multi-selection requires some deeper structure change, so really that work needs to happen first, before I can make it happen. This is why I delayed it for some time where I will be in vacations .. :-)

Thank you for the suggestions, I will look at implementing that than without the "a" probably indeed to make them shorter ..).

Btw, 2.0.54 will be out tonight.

**Comment by elgreko5** [comment]
Here's the video - bsp2-sidebar-dragging.ts - https://ufile.io/i3lok

**Comment by aaFn** [comment]
Thank you.
So by looking at the video in details, all seems normal:
- First time you drag, you do not go too deep, so the horizontal scrollbar at bottom remains hidden, and nothing happens.
- Second time and further sweeps, you go deeper inside, so the horizontal scrollbar is re-enabled.
- Last attempts your mouse is moving quicker, so no event is sent to the add-on by FF before you reach the auto-scroll area, and the horizontal scrollbar doesn't even disappear.

=> when you are on the auto-scroll area that FF controls, and the scrollbar is enabled, auto-scrolling on drag occurs. That is a normal FF and HTML feature.
Remain there and it will scroll to end or until you exit the auto-scroll area, either by the left (deeper inside the sidebar), or by the right.

This is normal and wishable as I said previously, in case for example you are on a tree with lots of levels and you need to scroll right to make visible the place where you want to drop the object you are dragging.

**Comment by elgreko5** [comment]
Ok then, if you wanted it to work that way, but methought that scrolling should only start when it is needed - and that is not only when there is long-named item in visible sidebar window, but when also the pointer is on an item with too long a name for a window.
As it is now, you get unwanted scrolling when you change your mind and decide not to bookmark, just because there is long named item somewhere outside of visible sidebar, and you need to scroll it back after you drop the tab back, and if you stay at auto-scroll area and name of not visible item is long enough, you'll eventually get blank sidebar window.

**Comment by aaFn** [comment]
Yes, this is the way auto-scroll functions in FF = the zone triggering auto-scroll is a rectangular band, aligned with the border it is against and having its l height (or width) -> form cannot be controlled, size cannot be controlled / changed, the fact it is active / inactive cannot be changed .. etc ..

=> As soon as you have a scrollbar, there are two auto-scroll zones perpendicular to it, one on each end, and that doesn't depend on the object your pointer is on.

Since they are needed for when we want auto-scroll, then we have to compose with that effect .. I found a way to inhibit it when entering the sidebar, but have to restore it at point, which is when you go deeper in the sidebar (you are not supposed to re-exit once you entered sufficiently deep :-) )

**Comment by elgreko5** [comment]
Hello aaFn,

Found another little bug: :)
Open BSP2 in new tab - right click on an item - when over text cursor is not an arrow

**Comment by aaFn** [comment]
Hello @elgreko5, correct. Not sure exactly why it is different in a tab, but fixed in 2.0.55 to come, by forcing the default pointer everywhere on menus.

Thank you, aaFn.

**Comment by elgreko5** [comment]
We talked about fonts - please, try this on a W7 machine when you catch time:

- Open BSP2 in a tab
- Open (FF original) Bookmarks 
- In BSP2 options check "Override default bookmark item spacing to" and set it at 6
- In FF Options - Fonts & Colors - Advanced, uncheck "Allow pages to choose their own fonts", press OK, then change Default font to Arial
- Look at both bookmarks - now BSP2 bookmarks look much more like original and are more legible

You wrote that it isn't possible for add-on to discern between W10 and W7 but that isn't necessary.
The solution is to add new option for changing font where default would be Tahoma which could be changed to Arial.

**Comment by elgreko5** [comment]
This BSP2 in a tab is great! Bravo!

You just need to squash a few bugs:
1. Drag&drop-ing bookmark from a tabbed BSP2 doesn't work.
2. When drag&drop-ed from a tabbed BSP2 to FF Bookmarks, in copied bookmark field Name gets replace with contents of field Address

This is the next best thing since show&go parent folder! :)

**Comment by elgreko5** [comment]
You wrote: I added a context menu on the magnifier glass and on any empty part of the sidebar body, to propose a context menu with an action to open BSP2 in a tab.

It's not working on any empty part (e.g. doesn't work south of "Other bookmarks") but only in a rather narrow frame around search rectangle.
Also, invoking it through search icon is counter-intuitive.

If possible, it would be better to put that command in a BSP2 menu.

**Comment by elgreko5** [comment]
Another suggestion:
Item Properties' box is too small - should be two times wider (which would also mostly solve that it doesn't look good when path is too long for one line)
Also, the box first appear for a moment in a sidebar which it shouldn't (Maybe a shortcoming of Mozzila extensions?)

**Comment by aaFn** [comment]
Hello @elgreko5,

yes, looking at making drag and drops function between sidebar and tab, and native bookmark sidebar and tab. That was not the intent initially, but now that it is there, and not too hard to get working, let's make them operational.

For context menu, the area south of "Other bookmarks" is not a blank part .. this is the bookmark pane, so you are correct, there is not a lot of space. Hence the idea to set it also on the magnifying glass, as this one is easier to click on than the rest of blank space ...

Not sure what you call a BSP2 menu ? Do you mean upon right click on the yellow star in the toolbar ?

Properties box = remember, this is a browser window .. you can resize it to whatever size you want.
If you want it to keep its size once you changed it, simply set "Remember size of resieable objects ..." in the add-on Options page. .. and you can use the `Reset defaults` button when you want to get back to initial size.

Yes, the property window visually moves ... that again a bug in FF .. BSP2 retains the last position of the property window when it is closed, and gives back thes coordinates to FF when opening it, however FF simply ignores them .. so workaround is to create the window at random (as FF decides) and to move it afterwards where needed ... :-(

**Comment by elgreko5** [comment]
> Not sure what you call a BSP2 menu ? Do you mean upon right click on the yellow star in the toolbar ?

I wasn't clear - I meant menu above search box

**Comment by aaFn** [comment]
Ok, for now, I prefer not to charge more the interface with such a thing.
Let's say it is an advanced function ..

I am progressing to support drag & drop between tabbed version and sidebar version, and also with native Bookmark sidebar

.. BUT .. (there is always a "but", did you notice ?)

- I cannot make a folder drag work specifically from BSP2 to the native Bookmark sidebar, it ends with error "_Can't copy a container from a legacy-transactions build_".
Indeed, I got no way for me to retrieve the browser instanceId, cf. https://bugzilla.mozilla.org/show_bug.cgi?id=1386513 and https://bugzilla.mozilla.org/show_bug.cgi?id=1177654
- Drag of a bookmark or of a separator specifically from BSP2 to the native Bookmark sidebar ends up in a copy instead of a move, with the following message in the browser console = "_Tried to move an unmovable Places node, reverting to a copy operation._" ..
Again apparently for the same reason that I cannot retrieve the browser intanceId and give it in the drag dataTransfer.

Hope this will be ok, that appears to be as much as I can go for now ...

**Comment by aaFn** [comment]
I opened a bug to FF to see if we can have a way to get that missing value ..
-> https://bugzilla.mozilla.org/show_bug.cgi?id=1531539

**Comment by aaFn** [comment]
.. and 2.0.55 ois out with the drag and drop additions / corrections.
Let me know.

**Comment by elgreko5** [comment]
Hello aaFn,

> Ok, for now, I prefer not to charge more the interface with such a thing.

Don't understand, but doesn't matter...

On to more bugs:

1. D&D on an empty folder doesn't work.

2. Also, sometimes dropped bookmark doesn't land in a folder but between folders. And it does several times in a row, and then it works as it should.

3. When even after seeing my video you still claimed that D&D scrolling bug is normal and wishable, I gave up.
But now I see a new chance for you to understand - try moving bookmark from sidebar to tab. If you still think it's normal and wishable, I'll be done with trying.

4. That light blue highlight when hovering over a bookmark is barely visible.

**Comment by elgreko5** [comment]
Another one:

5. D&D of a bookmark that sits between folders in a sidebar, to some folder, doesn't work.

**Comment by elgreko5** [comment]
6. Sometimes, when D&Ding a tab to sidebar to bookmark it, when flying over to drop it on a folder, blue indicator line rests on a random place (between bookmarks). Blue line survives scrolling up and down out of a visible window and back. This bug is not introduced in newest version, it was there before.

**Comment by aaFn** [comment]
Hello @elgreko5 

1. Looks like a regression, not sure when it was introduced, I will look at it and correct.
Sorry.

2. When you drag, there is a blue insertion line going with it, or a folder highlight.
If folder highlight, the bookmark will be put in the folder, at end of it.
If blue insertion line, the bookmark will be inserted at that place.
You need to look at that to decide what you want to do (insert into, or between).

3. The problem to solve is to distinguish between two cases, which are exactly identical until the drag gets out of the sidebar ... but so far there is no way to know in advance what is your intention:
   - Say you want to move a bookmark from high in the tree (=> left of the sidebar) to very deep in the tree (=> to a place on the right not visible in the sidebar).
=> auto-scroll is needed, to bring the right part into view, and drop the bookmark there = this is whishable.
   - Take the case you mention = all goes the same until you get out of the sidebar b ythe right, at which time we find that the auto-scroll was not useful and not desirable.
   - But should the add-on know in advance that you are in first or second one ?
And by the way, you could well be in first case, but given you went too fast with your mouse, you exit by the right, just to come back later .. and so the auto-scroll was still opportune and wanted.
   - One way could be to press a key at same time to tell BSP2 to disable the horizontal scrollbar, and that would disbale auto-scroll ... but this is non natural and not obvious ... isn't it ?
   - Listening to other suggestions ...

4. That highlight color was the one of FF at the time I created the add-on. Now FF uses a light grey .. not sure this is any better.
I can increase the color saturation if you believe that would be better, let me know.

5. Looks like the same case than 1) .. I will look to fix this week-end, again sorry for the regression.

6. Correct, I saw that on Linux, and didn't have the time to investigate why, but I suspect this is because in some cases FF doesn't send the "exit" or "leave" event to the add-on ...
I'll what I can do about it ..
I am not able to reproduce it in Windows ... at least not on my machine.
Do you have a particular gesture which makes it 100% happen ? That will help investigate and devise a workaround ...

**Comment by elgreko5** [comment]
3. It's simple and nothing new - even FF uses it: hovering for a short time (longer than just flying over) above triggering zone.

**Comment by elgreko5** [comment]
Sorry, forgot to add that bug 1 (D&D on an empty folder doesn't work) manifested itself when D&Ding from tab to sidebar. Because of that I reported 5 as another bug. Also, D&D from tabs to an empty folder works.

2. You can't seriously think that after bookmarking 80.000 items I don't know that.
Please, reread what I wrote: "2. Also, sometimes dropped bookmark doesn't land in a folder but between folders. And it does several times in a row, and then it works as it should.".

4. 
![11](https://user-images.githubusercontent.com/47822770/53651335-babf7e80-3c46-11e9-9e32-591d3c249d82.png)
![12](https://user-images.githubusercontent.com/47822770/53651345-c14df600-3c46-11e9-9978-addda7d22acc.png)

Light blue and light grey - it, of course, depends mostly of luminosity of used color. Maybe #D7EDFF would be better than #E5F3FF you're currently using.

Also, see how it looks on W7 - when pointed at, bookmark's name is underscored.
![13](https://user-images.githubusercontent.com/47822770/53655646-10992400-3c51-11e9-9281-0ceb10ebca19.png)

**Comment by aaFn** [comment]
Hello again,

1. I found the regression, I just introduced it in 2.0.55 when modifying the D&D operation to support dragging from native bookmark sidebar into tabbed BSP2. Thank you for noticing that fast.
Corrected in 2.0.56.

2. Sorry, I must be slow, but I am not sure I understand what you mean. Now it might be an effect of bug in 1) in fact, because while I was investigating it, I did see random results on the drop location when related to folders.
Now that 2.0.56 is out, can you try again with it and let me know ?

3. Once again, I do not have control over auto-scrolling, nor on where is its triggering zone (in fact I do not even know where is this zone), nor after how much time it is triggered (I believe it is 0s by the way from observation).
And it is not an "object" I am notified about when the mouse pointer is hovering on it.
This is completely controlled by FF. All I get is a scroll event when it was triggered, but too late, the scroll already happened.
I can only disable or enable the horizontal scrollbar in advance, but this is much before the fact.
And as I explained, I do not know in advance, when you start dragging, if you want to auto-scroll to right, or if you want to go outside, so cannot decide if the scrollbar should be disabled or not.
However, if I disable it all times, I am preventing the first use case, while if I leave it on, the impact is annoying for second use case, but it is not a limitation, both use cases are allowed.
Therefore, I do not disable the scrollbar, this is the lesser evil.

4. Remember that FF changed colors since I started BSP2. The colors I have correspond to FF57.
In windows 10, FF light grey is now ECECEC .. definitely too light.
I will play with the color you suggest, and also saturate more the selection .. probably to match the new blue of FF .. Will be for 2.0.57 as I want 2.0.56 to focus on fixing the bugs.

5. See 1)

6. I believe I found a paliative, let me know if there remains some cases on your side.

**Comment by elgreko5** [comment]
2.0.56

DD from BSP2 tab to BSP2 sidebar moves bookmark.
DD from BSP2 tab to FF sidebar copies bookmark.

I found out what's going on with seemingly random placing of a bookmark between the folders:
DD from FF (or BSP2) sidebar to first place in a folder in BSP2 tab moves bookmark, but places it between that and next folder. It's not random but happens when the bookmark is dropped on a first blue line - the problem is that you have two blue lines between folder and first bookmark in that folder. Second blue line is OK, first should be eliminated.

**Comment by aaFn** [comment]
Correct, when dropping in the line just after a folder, this drops it after the folder **at same level**, not inside it.
This behavior is consistent, as it is the same whether the folder is open or closed. It allows to add an item at end of the bookmark pane, when the last item is a folder (a thing not possible in the native FF sidebar, unless the folder is closed .. and still, difficult to catch),

And correct, drag from BSP2 to FF sidebar copies the bookmark or separator, as explained above, because of https://bugzilla.mozilla.org/show_bug.cgi?id=1531539 (and results in nothing happening for a folder, because of same reason).

**Comment by aaFn** [comment]
On colors, by the way, note that the light blue is exactly the same as Windows 10 windows explorer, when hovering on items.

**Comment by elgreko5** [comment]
>It allows to add an item at end of the bookmark pane, when the last item is a folder whether it is open or closed (a thing not possible in the native FF sidebar, unless the folder is closed .. and still, difficult to catch),

I reread that I still don't understand? 

Also, I now see there are also 2 lines between 2 closed folders - dropping on either of the lines has the same effect?

**Comment by aaFn** [comment]
Each item, whether folder, bookmark or separator, has 2 insertion lines: 
- 1 above (= insert before as sibling / brother)
- 1 below (= insert after as sibling / brother)

That means indeed that when inserting between two items (which is the majority of cases, except then appending), there are two lines.
What happens is as follows:
- if the two items are at the same level (= same parent), that makes no difference, "after item 1" == "before item 2"
- however, if they are not at the same level, e.g. one of the items is the last element of a parent folder, and the next item is the next sibling / brother of that parent folder, then
  - the line for insert after item 1 will insert as new last element inside parent folder
  - the line for oinsert before item 2 will insert at same level than parent folder and item 2, and so won't be in parent folder
  - difference can be seen by looking at where the insert line is starting, second line is more on the left than first line in the case I described.

Hope this is clearer, but try and you will experience by yourself, it is better than words.

**Comment by elgreko5** [comment]
As I wrote:
>It allows to add an item at end of the bookmark pane, when the last item is a folder whether it is open or closed (a thing not possible in the native FF sidebar, unless the folder is closed .. and still, difficult to catch),

I reread that and I still don't understand? 

That means I don't understand what's quoted, not what you wrote about 2 lines.

**Comment by elgreko5** [comment]
>Hope this is clearer, but try and you will experience by yourself, it is better than words.

It wasn't cloudy at all, except that I thought of it as a bug and not a feature :)

>This behavior is consistent, as it is the same whether the folder is open or closed.

OK, there is, of course, some logic in just everything, as well as everything is consistent with itself.
Everything has it's upsides as well as downsides.
Upside, you mentioned one (which I don't understand).

Downsides are many and IMPORTANT.
It is not consistent with simple, intuitive notion (that we so far had) that bookmark should land where blue line points at! Now, we'll have to think about and to look up to see if we are in FF or BSP2 bookmarks, so as to accordingly change our moves!
I thought it was a bug because the same moves I was doing earlier now had different effects. Now, after I discovered what was going on (and you explained it was a feature), I MUST be more precise in those moves, and to be more precise I must SLOW DOWN considerably.
I think I must emphasize it's not just my point of view - every human must slow down to be more precise!
And to be more precise, one must focus, focus away from a job one's doing. And then, focus back.

It should always be on programmer's mind - the best interface is no interface (also, title of a book :) ).

**Comment by aaFn** [comment]
Hello @elgreko5, ok your arguments of not landing where the blue line is, and of becoming more precise so slower, are good, you convinced me
=> I will change the behavior in 2.0.57.

Thanks, aaFn.

**Comment by elgreko5** [comment]
Hello aaFn,

Another bug:

Sometimes a folder doesn't close on first click. I've seen this earlier but thought that maybe mouse switch was acting up. Can't define exact circumstances.

**Comment by elgreko5** [comment]
And another one, happened just now:

When I tried to catch scrollbar thumb I couldn't. Although arrow was above thumb, BSP2 thought it wasn't. Holding left mouse button down while on a track below thumb, resulted in scrolling but it didn't end where it should - the thumb didn't stop directly below arrow's tip. After some scrolling up and down, it was again possible to catch a thumb with left mouse button, but holding left mouse button on a track still doesn't do what it should.

Also, thumb's minimal height is too small - should be bigger - FF's height is OK (IIRC they made it bigger recently).

**Comment by elgreko5** [comment]
And now, for something completely different :)

Today I did with bookmarks what I usually do with them - edited bookmarks' names, made a folder, 10 bookmarks in it and edited them.
Saw one bookmark I made earlier in Other Bookmarks and tried to move it as well as delete several Mozilla's bookmarks that were there. Deleting them had no effect. Tried to delete some folder - the same.
Closed FF only to see it crash.
After reopening FF the folder I made earlier was gone - together with it's 10 bookmarks!

Switched to BSP2 and folder was there! Wasn't taking chances copying folder to FF Bookmarks, so opened them in new tabs and bookmarked them from there (to FF B.). I lost what I edited (names) but BSP2 at least saved my bookmarks. Of course, after I switched to FF and again to BSP2 sidebar, BSP2 synchronized, erasing them. (Hm, it seems not - now in BSP2 I have original folder and folder that BSP2 took from FF's bookmarks!?)

I presume this behaviour of BSP2 wasn't intended but it is highly desirable so I urge you to watch out not to break it in future versions, and also to consider adding an option to save BSP2 bookmarks (for such a case) in a file identical to FF's so it could be manually replaced. Current functionality isn't enough except in above mentioned special case when bookmarks are in one folder.

Also, I think you should state described feature in add-on's description (as well as way to use it - only immediately after reopening FF, before BSP2 synchronizes), so users would know about it.

**Comment by aaFn** [comment]
Hello @elgreko5 ,

1. Folders not closing on fist click -> are you sure you are not trying to close an already closed folder ?
Let me explain:
    - When you do a search, then click on a search result, this one is shown in the bookmark pane
    - As per it was done in BSP1, BSP2 is expanding closed folders for visibility, **but it does not open them**
    - This is so that when you stop the search (press ESC or on the cross at end of searchbox), all the tree is restored as it was before, not modifying your tree state
    - In such a case, the twistie at left of the folder is still a closed one, so when you click on it, you are setting the folder to "open" state = visually, nothing is changing except the twistie which becomes "open", however, when you will stop the search, the folder will now remain open, because you modified the tree state.
    - To close it back, you have to click a second time on it, to turn the twistie to "closed"

2. Scrollbar strange behavior = this is all controlled by FF, nothing is done by the add-on on that object, so the blame should go on FF :-(

3. Scrolbar thumb size = nothing the add-on can change here. There are only 2 things that can be changed in FF in HTML/CSS, and they are not the thumb length, only the scrollbar color and width -> https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Scrollbars
As said previously, I believe the native bookmark is not really using HTML objects, but native OS objects, or a web layer closer to the OS, so it can do things that add-ons cannot do.

4. BSP2 seeing things that the native bookmark sidebar does not anymore -> yes, already observed with other users here who have high numbers of bookmarks.
    - It seems there can be some corruption in the FF bookmark DB, don't know what is causing that
    - Another user was telling me that from time to time, entire subtrees were lost by the native sidebar (still there, but not shown), and in such cases he was using BSP2 to retrieve them and bring them back to life by rehooking them / moving them
    - I suspect that this is due to the large query that BSP2 makes to FF API = basically, no filter applied, no parent .. etc .. so all is returned
    - Then BSP2 simply walks down the tree, children by children, to rebuild in its memory the full tree
    - It appears that this children relationship is not the corruptable piece in the FF bookmarks DB, hence the feature of retrieving them.
    - Not an intended feature, this is simply relying on that general "all bookimarks" query
    - I do not intend to change that
    - However, FF might change it .. and then I can't do anything about it .. so can't guarantee it will always be like that.
    - To reassure you, I believe though there are very few chances that FF decides to change that :-)

5. Export to file = not possible, an add-on is not allowed to my knowledge to read from or to write to OS files on disk.
    - What you can do though, as the all mighty user in FF :-), is the following:
    - Press on Shift, then right click inside BSP2 sidebar -> a menu that no add-on nor Web page can control will appear
    - It has a "Save as..." action
    - Use it, and it will create a panl.html + linked folder (or whatever name you choose to give to it)
    - In panel.html, you will retrieve all your bookmarks, even the closed / non visble ones, in HTML form in a `<table>` element.

Except from the above last procedure, one thing I could do would be for me to create a new tab with the text in it that looks like a bookmark file exported by FF.
Then you would still have to use the FF File menu, or shift+right click, to save it as a file on your PC where you want ... => if this is what you want, can you create a separate issue to track that specific demand ?

Thanks, aaFn.

**Comment by aaFn** [comment]
2.0.57 is out ..

**Comment by elgreko5** [comment]
Installed 2.0.57, thanks.
Hate to say it, but you didn't change insertion as last item - there are still two lines...

1. It might be that, I seem to have seen twistie changing. Will pay attention.
Regarding preserving previous tree state, I don't see that particularly useful, if at all. In fact, I think it would be far more useful to be able to go to particular bookmark/folder through the search, particularly as we now have BSP2 in a tab.

While with preserving - after ESC, the view in sidebar is not restored to what it was?

Also, is there a particular reason you have changed color of twistie from FF's? I suggest you revert to FF's color - it is more legible and now twistie drowns in that gray highlight (in BSP2).

Having said that, I commend changing that FF's idiotic inverted highlight to normal one. That mixing of normal and inverted text is a sign of ignorance of ergonomics.

2. Then it's another FF bug, and idiotic one - it works correctly in FF sidebar.

3. See 2.

4. Well, if for FF developers 80.000 bookmarks is a big database, they should change profession. Application that is losing data from it's database is not application - it is just garbage.

What I am more and more assured of is that I should be moving away from FF, ASAP. There is a saying: you can't make a pie from sh*t.

5. Tried that Save Page As... - thought FF stopped responding so I killed it. Second try - almost killed it again - took 10 MINUTES on i5?! Ridiculous. 
I understand it is FF developers' art.

What I meant was any way to save a file with which could FF's bookmarks file be replaced (or imported). Of course, if you're willing and it is not too much work for you.
Also, could you advise why is Library/Import and Backup/Backup producing a file of 38 MB but Restore option is showing FF's automatically generated files with lengths of 13.9 MB?

Coincidentally, after installing 2.0.57 I noticed that duplicated folder disappeared from BSP2 too... That probably means FF first marked them as deleted and afterwards really deleted them.

Please, consider adding counters of bookmarks and folders somewhere (e.g. instead of "v2.0.57" in a heading of a BSP2 menu). That could really help cutting the losses with that most serious bug.

**Comment by aaFn** [comment]
Not sure where you see two lines on an open folder with contents ?? I just retested to make sure, it's not there at all. Only on closed folders, or empty folders, but that is normal since there is no ambiguity in such cases, and the 2 lines play their full role there.

1. Correct, after ESC, the view is restored to what it was (except what  you modify). This is BSP1 behavior, so I kept it by default.
If you don't like that, simply tick the option "After show bokmark, leave bookmark tree open ...", and the tree will remain as it is from opening to the path to show the bookmark . but then you will spend your time closing things manually later when you do not need to see it anymore ... your choice :-)

The twistie color and form are the ones from FF 57. This is FF who changed .. not BSP2 .. Frankly, I prefer the one of FF57, so I didn't change ..

Differences in sizes in FF backup and restore .. absolutely no idea .. one more mystery of life :-)

Counters of bookmarks = can that be ok with Stastistics in the Options page for now ?
The version (e.g. v2.0.57) was a demand of a number of users, so I'd rather not remove it now.

I'll see for creating an export page if you want, but I believe the Library -> Import and Backup and then Backup or Export to HTML are complete, even if the native sidebar does not show all, so you can safely use them.
I say that because one of the users who has problems like you of disappearing branches explained to me this is how he was in the past retrieving his "lost branches", before now relying on BSP2 to do that.

**Comment by elgreko5** [comment]
Yes, they are :) - open BSP2 in a tab - open one folder - drag bookmark from sidebar past last item in a folder...

As I wrote - after ESC, the view in sidebar is NOT restored to what it was - try and see.

Look at that twistie drowning itself in that gray highlight - some of twistie's colors are the same as highlight's gray?! Maybe your monitor is better than mine - Eizo IPS pro monitor with 4.000-10.000 K color temperature, gamma, saturation, hue, overdrive, gain, whatnot, but no monitor shows the same colors as different ones :)

I know you put counters there but the purpose of putting them where I suggested (and be updated in real-time) was to be able to see when numbers change in time to be able to save what FF ate. Maybe not so good idea.

Really surprised by that demand, different strokes for different folks but I can't possibly imagine ANY use for that? (Before this idea I even thought of suggesting that you remove version number from there, but thought you'll think I'm nitpicking. :) )

Regarding export - yes, it could be unnecessary, will try.

**Comment by aaFn** [comment]
Remember your initial point was about inserting as first position within a folder, where the first line was not physically representing where the insertion point really was ... that's the one that I agreed to remove. Other places, this is ok, they always represent where the bookmark will be inserted.

Having both is also what is in the native sidebar = there is an "insert before" and an "insert after" line for each item .. simply they do not showq the difference so you get a "random" effect without knowing.

Try the following in the native sidebar to observe that same phenomenon = open a folder, drag an external item towards the last element.
- Drop it close, but below, the center of the last element = it will insert as next sibling under same falder, becoming the new last one
- Drop it close, but above, the center of next visual element which is at a higher level in the tree = it will insert as previous sibling of that next element, and not under the open folder.

You have the same insert before and after thing in BSP2, the only difference is that in BSP2, the line horizontally changes and adapts to the sie and position of the element it relates to, in order to show the difference, while in FF native it doesn't ..
Looks better to me as then you know what will happen, while in native you can only rely on your estimate of how close you are to the center of one or the other to anticipate / drive what will happen ...

**Comment by elgreko5** [comment]
You skipped that about drowning twistie?

Initially, yes, but I also later wrote: "Also, I now see there are also 2 lines between 2 closed folders - dropping on either of the lines has the same effect?"
It is a bug, and it's still the case.
adjacent
For the other cases, cases when adjacent items are ON DIFFERENT LEVELS, you're right, but to put item in last place I just drop item on a folder. I think that moves requiring precision of several pixels should be avoided whenever possible.

Another bug:
Context menu on item in BSP2 sidebar doesn't disappear on click on thumb.

Another thing:
In context menu on an item in search results there are "Show bookmark" and "Go parent folder". Different words, "show" and "go", denote the same action. Also, Show bookmark is redundant.

**Comment by aaFn** [comment]
Twistie and colors = I will see if I have time to improve

2 lines = this is like native as I explained, and even better because the changing line shows really were you are, so I won't change that. If you consider that as a problem, then you have the same problem on native.

Context menu = that is not a bug, but an FF thing .. BSP2 does not receive any event when you click on the thumb, so BSP2 simply doesn't know (and so nothing will happen in it therefore)

2 actions = they are different .. as their names say, "Show bookmark" highlights the bookmark, "Go parent folder" highlights the parent of that bookmark. And in some cases, that parent is very far from the bookmark (some users have folders with 100's of bookmarks).

**Comment by andreygursky** [comment]
> The value I am at today is as far as i could get today with using FF API on local storage (which is far from efficient, it is storing json text, not binary, and it is slooow and hungry in memory..) and on the HTML display speed on FF, and I believe I got all the juice I could get from it.

It looks like there are good news after recent Firefox 66 release: https://blog.mozilla.org/addons/2019/02/15/extensions-in-firefox-66/
> I want to start by highlighting an important change that has a major, positive impact for Firefox users. Starting in release 66, extensions use IndexedDB as the backend for local storage instead of a JSON file. This results in a significant performance improvement for many extensions, while simultaneously reducing the amount of memory that Firefox uses.
>
> This change is completely transparent to extension developers – you do not need to do anything to take advantage of this improvement.  When users upgrade to Firefox 66, the local storage JSON file is silently migrated to IndexedDB. All extensions using the storage.local() API immediately realize the benefits, especially if they store small changes to large structures, as is true for ad-blockers, the most common and popular type of extension used in Firefox.

**Comment by aaFn** [comment]
Indeed, let's see for you guys who have plenty of bookmarks if this is giving you better results ... :-)
Let me know.

Quite interesting is also this one: https://bugzilla.mozilla.org/show_bug.cgi?id=1419195
It means that now, with FF66+, when you are in the Bookmarks Library, you can right click on a bookmark .. and you see the BSP2 context menu giving you the path to the bookmark :-)
Not as seamless as another column, as it was with alice0775's Show Parent Folder, but still there and easily accessible nonetheless ...

**Comment by aaFn** [comment]
@elgreko5 : Twistie and colors = I am improving them in 2.0.58, published this evening.
Closing this thread now, it is long enough, and quite complex :-)

If you have other points, it is best to open other specific issues, one for each of them.

Thank you, aaFn.
